packages/modern-browsers/modern.js

```diff
@@ -0,0 +1,123 @@
+const minimumVersions = Object.create(null);
+const hasOwn = Object.prototype.hasOwnProperty;
+
+// TODO Should it be possible for callers to setMinimumBrowserVersions to
+// forbid any version of a particular browser?
+
+// Given a { name, major, minor, patch } object like the one provided by
+// webapp via request.browser, return true if that browser qualifies as
+// "modern" according to all requested version constraints.
+function isModern(browser) {
+  return browser &&
+    typeof browser.name === "string" &&
+    hasOwn.call(minimumVersions, browser.name) &&
+    greaterThanOrEqualTo([
+      ~~browser.major,
+      ~~browser.minor,
+      ~~browser.patch,
+    ], minimumVersions[browser.name]);
+}
+
+// Any package that depends on the modern-browsers package can call this
+// function to communicate its expectations for the minimum browser
+// versions that qualify as "modern." The final decision between
+// web.browser.legacy and web.browser will be based on the maximum of all
+// requested minimum versions for each browser.
+function setMinimumBrowserVersions(versions) {
+  Object.keys(versions).forEach(browserName => {
+    const newMinVersion = versions[browserName];
+    if (hasOwn.call(minimumVersions, browserName)) {
+      if (greaterThan(newMinVersion, minimumVersions[browserName])) {
+        minimumVersions[browserName] = copy(newMinVersion);
+      }
+    } else {
+      minimumVersions[browserName] = copy(newMinVersion);
+    }
+  });
+}
+
+Object.assign(exports, {
+  isModern,
+  setMinimumBrowserVersions,
+});
+
+// For making defensive copies of [major, minor, ...] version arrays, so
+// they don't change unexpectedly.
+function copy(version) {
+  if (typeof version === "number") {
+    return version;
+  }
+
+  if (Array.isArray(version)) {
+    return version.map(copy);
+  }
+
+  return version;
+}
+
+function greaterThanOrEqualTo(a, b) {
+  return ! greaterThan(b, a);
+}
+
+function greaterThan(a, b) {
+  const as = (typeof a === "number") ? [a] : a;
+  const bs = (typeof b === "number") ? [b] : b;
+  const maxLen = Math.max(as.length, bs.length);
+
+  for (let i = 0; i < maxLen; ++i) {
+    a = (i < as.length) ? as[i] : 0;
+    b = (i < bs.length) ? bs[i] : 0;
+
+    if (a > b) {
+      return true;
+    }
+
+    if (a < b) {
+      return false;
+    }
+  }
+
+  return false;
+}
+
+// ECMAScript 2015 Classes
+setMinimumBrowserVersions({
```

Comment:
We're exporting `setMinimumBrowserVersions` and allowing it to be called multiple times as needed which is great, but once all of the `setMinimumBrowserVersions` calls have been completed and we have the new definitive `minimumVersions` ready to go, what happens if at some point we notice the `minimumVersions` aren't quite what we expected? How do we track down what call to `setMinimumBrowserVersions` is impacting the minimum versions used, especially since `setMinimumBrowserVersions` is part of the public API and can be called by 3rd party packages? The current values added via `setMinimumBrowserVersions` don't have a feature identifier of any kind associated with them, so we don't have an easy programmatic way to track down what feature is causing the minimum versions to be set at a certain level. We have to either look at the source, track down inline comments (like `ECMAScript 2015 Generator Functions`, `ECMAScript 2015 Template Literals`), or rely on other forms of documentation. If the `setMinimumBrowserVersions` are spread across multiple packages, this could be a pain. I'm  wondering if we could somehow mark a certain minimum browser version set with an identifier, and track that identifier when setting the `minimumVersions` baseline. So instead of

```js
// ECMAScript 2015 Generator Functions
setMinimumBrowserVersions({
  chrome: 39,
  edge: 13,
  firefox: 26,
  mobile_safari: 10,
  opera: 26,
  safari: 10,
  // Disallow any version of PhantomJS.
  phantomjs: Infinity,
});
```

we have something like

```js
setMinimumBrowserVersions('ECMAScript 2015 Generator Functions', {
  chrome: 39,
  edge: 13,
  firefox: 26,
  mobile_safari: 10,
  opera: 26,
  safari: 10,
  // Disallow any version of PhantomJS.
  phantomjs: Infinity,
});
```

then when deciding on the `minimumVerions`, we track the identifier for the version that ended up taking priority and being set. So anyone with programmatic access to `minimumVersions` could dump it out and see something like:

```js
{
  chrome: {
    decidedBy: 'ECMAScript 2015 Classes',
    version: 49,
  },
  edge: {
    decidedBy: 'ECMAScript 2015 Generator Functions',
    version: 13,
  },
  firefox: {
    decidedBy: 'ECMAScript 2015 Classes',
    version: 45,
  },
  ...
}
```

Then again this might be over engineering ...

Reply:
I decided to do this by examining the call stack rather than passing in a string: 93c263225250c4431c4b0cb82b509b3eb07bcb0c. What do you think?

Reply:
Even better idea @benjamn - this looks great!